### PR TITLE
redis: add support for redis modules with custom commands

### PR DESCRIPTION
**Pull request by inf-rno**
Commit Message:
- Fixes https://github.com/envoyproxy/envoy/issues/12822
- Make SupportedCommands a Class with the built in commands as default
- Add additional commands for the different command types via a listener config
- Pass the supported_commands_ around to use in place of the global static
Risk Level: Medium
Testing: 
Docs Changes:
Release Notes:

### CODE REVIEW

**Comment by inf-rno** [comment]
First stab at adding modules support. I havent fixed the tests yet but wanted to make sure that this idea is ok. cc @mattklein123 @FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg @HenryYYang

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
It seems reasonable to categorize module command behavior into our established types of commands; is that flexible enough for most redis modules? I have not used modules before. Any advantages to just having a custom commands section, vs a modules with commands section?

3 big considerations with the redis code in Envoy: 

1) Redis cluster vs regular Redis 

2) Mirroring traffic. If your module commands behave exactly like the established commands, things should work ok. I advise setting up two local redii (vanilla, non-cluster) and checking if mirroring works, and then trying the same with two redii in cluster mode (using docker compose or the like makes it less painful), and making sure commands end up where you expect them.

3) MGET behavior- MGET is decomposed by envoy into GET: Code: https://github.com/envoyproxy/envoy/blob/02e97e1dcd948c14b62de95b348badb5ee6d1575/source/extensions/filters/network/redis_proxy/command_splitter_impl.cc#L240
Test: https://github.com/envoyproxy/envoy/blob/02e97e1dcd948c14b62de95b348badb5ee6d1575/test/extensions/filters/network/redis_proxy/command_splitter_impl_test.cc#L551
MGET style behavior essentially requires a custom implementation each time- perhaps it's time to generalize that @HenryYYang ? This might be helpful when we get around to implementing a scan that works in cluster mode.

**Comment by inf-rno** [comment]
The command splitter in envoy mainly cares about 
- where the keys is: simple or eval style
- read or write
- optimization with hash sum style for fanout purposes
- mget/mset

As long as modules fit within these, they should map well. But otherwise I don't see how envoy can support them. For our use case (CAS operations similar to memcached) we have simple commands and mget/mset style commands. So this should work. 

I'll clean up the tests and update the PR.

**Comment by inf-rno** [comment]
Sorry this fell of my plate for a bit there. I have updated all the tests and this is ready for review.

**Comment by mattklein123** [comment]
Please check CI

/wait

**Comment by mattklein123** [comment]
Merge main and check CI? I can take a look then.

/wait

**Comment by inf-rno** [comment]
CI seems to be failing on format checks that got introduced by `check_format.py`. I had to manually unstage these last time to help out CI. Is this expected? I am also not able to push with verify because the pre push hook gets stuck in a `less` output somewhere and quitting that terminates the push.

**Comment by mattklein123** [comment]
You need to fix format locally to get it to what CI expects.

**Comment by inf-rno** [comment]
I've had to push with `--no-verify` because otherwise I the pre-push hooks complain ```ERROR: check format failed. run 'tools/code_format/check_format.py fix'```. 🤞 on CI

**Comment by mattklein123** [comment]
Looks like more CI issues. Also friendly request to please don't force push as it messes up reviews.

/wait

File: api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto [hunk 1]
```diff
@@ -208,6 +208,28 @@ message RedisProxy {
     repeated string commands = 4;
   }
 
+  // RedisModule defines modules with custom commands.
+  // [#next-free-field: 7]
+  message RedisModule {
+    // Name of the module.
+    string name = 1;
```
Comment (mattklein123) [inline api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto:215]: Should this have a required / min length validation?

File: source/extensions/clusters/redis/redis_cluster_lb.h [hunk 2]
```diff
@@ -95,10 +97,10 @@ class RedisLoadBalancerContextImpl : public RedisLoadBalancerContext,
 
 private:
   absl::string_view hashtag(absl::string_view v, bool enabled);
-
-  static bool isReadRequest(const NetworkFilters::Common::Redis::RespValue& request);
+  bool isReadRequest(const NetworkFilters::Common::Redis::RespValue& request);
 
   const absl::optional<uint64_t> hash_key_;
+  NetworkFilters::Common::Redis::SupportedCommandsSharedPtr supported_commands_;
```
Comment (mattklein123) [inline source/extensions/clusters/redis/redis_cluster_lb.h:103]: nit: const

File: source/extensions/filters/network/common/redis/supported_commands.h [hunk 3]
```diff
@@ -4,89 +4,253 @@
...
-        "zremrangebyrank", "zremrangebyscore", "unlink");
-  }
+  static const std::string& ping() { CONSTRUCT_ON_FIRST_USE(std::string, "ping"); }
 
-  static bool isReadCommand(const std::string& command) {
-    return !writeCommands().contains(command);
+  bool isReadCommand(const std::string& command) { return !writeCommands().contains(command); }
+
+  void
+  addModules(const envoy::extensions::filters::network::redis_proxy::v3::RedisProxy& proto_config) {
```
Comment (mattklein123) [inline source/extensions/filters/network/common/redis/supported_commands.h:223]: Potentially some logging in here?

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.h [hunk 4]
```diff
@@ -178,6 +179,7 @@ class InstanceImpl : public Instance, public std::enable_shared_from_this<Instan
     Common::Redis::RedisCommandStatsSharedPtr redis_command_stats_;
     RedisClusterStats redis_cluster_stats_;
     const Extensions::Common::Redis::ClusterRefreshManagerSharedPtr refresh_manager_;
+    Common::Redis::SupportedCommandsSharedPtr supported_commands_;
```
Comment (mattklein123) [inline source/extensions/filters/network/redis_proxy/conn_pool_impl.h:182]: nit: const where you can, same elswwhere.

File: test/extensions/filters/network/redis_proxy/config_test.cc [hunk 5]
```diff
@@ -206,6 +206,39 @@ stat_prefix: foo
   cb(connection);
 }
 
+TEST(RedisProxyFilterConfigFactoryTest, RedisProxyModulesProto) {
```
Comment (mattklein123) [inline test/extensions/filters/network/redis_proxy/config_test.cc:209]: Do you have any tests that actually exercise that custom commands work? Can you potentially add an integration test of some type?

File: source/extensions/clusters/redis/redis_cluster.cc [hunk 6]
```diff
@@ -202,7 +202,8 @@ RedisCluster::RedisDiscoverySession::RedisDiscoverySession(
       client_factory_(client_factory), buffer_timeout_(0),
       redis_command_stats_(
           NetworkFilters::Common::Redis::RedisCommandStats::createRedisCommandStats(
-              parent_.info()->statsScope().symbolTable())) {}
+              parent_.info()->statsScope().symbolTable(),
+              std::make_shared<NetworkFilters::Common::Redis::SupportedCommands>())) {}
```
Comment (mattklein123) [inline source/extensions/clusters/redis/redis_cluster.cc:206]: Should this be some singleton that is created with CONSTRUCT_ON_FIRST_USE?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks overall LGTM. Flushing out a few comments. cc @FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg @HenryYYang can one of you take a quick look also? Thank you!

/wait

**Comment by inf-rno** [comment]
I'm sorry but this fell of my table the last month. We have taken a different direction to solve our need for CAS support in redis and as such we don't need envoy to support modules anymore. So I'm inclined to leave this PR closed as is and we can pick it up later if there is more interest. Sorry for the inconvenience.

**Pull request closed without merging**